### PR TITLE
test: circuit breaker for authz messages

### PR DESCRIPTION
**Pull request by rootulp**
## Description

Add a test to verify that the circuit breaker prevents authz transactions if the authz wraps a message type that isn't supported.

## Motivation

Motivated by this [security advisory](https://github.com/celestiaorg/celestia-app/security/advisories/GHSA-j4w8-q5hh-6xf9).

### CODE REVIEW

**Review by evan-forbes — DISMISSED** [review]

File: app/test/circuit_breaker_test.go [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+func getAddress(t *testing.T, account string, keyRing keyring.Keyring) sdk.AccAddress {
+	record, err := keyRing.Key(account)
+	require.NoError(t, err)
+
+	address, err := record.GetAddress()
+	require.NoError(t, err)
+
+	return address
+}
```
Comment (cmwaters) [inline app/test/circuit_breaker_test.go:102]: nit: I think we might already have a function for this in testnode

**Review by cmwaters — DISMISSED** [review]
Thanks for adding this test @rootulp!

File: app/test/circuit_breaker_test.go (hunk 1 above)
Comment (ninabarbakadze) [reply app/test/circuit_breaker_test.go:102]: yeah we do https://github.com/celestiaorg/celestia-app/blob/2e49d665b3275e769dfe0371b1ca41e39dc3f5f5/test/util/testfactory/common.go#L83

Comment (rootulp) [reply app/test/circuit_breaker_test.go]: https://github.com/celestiaorg/celestia-app/pull/3554/commits/69b10599cb18caa353203a3047260abfedb29bc2

**Review by cmwaters — APPROVED** [review]

**Review by evan-forbes — DISMISSED** [review]

File: app/test/circuit_breaker_test.go [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+import (
+	"testing"
+	"time"
+
+	"github.com/celestiaorg/celestia-app/v2/app"
+	"github.com/celestiaorg/celestia-app/v2/app/encoding"
+	v1 "github.com/celestiaorg/celestia-app/v2/pkg/appconsts/v1"
+	"github.com/celestiaorg/celestia-app/v2/pkg/user"
+	"github.com/celestiaorg/celestia-app/v2/test/util"
```
Comment (ninabarbakadze) [inline app/test/circuit_breaker_test.go:11]:
```suggestion
```

File: app/test/circuit_breaker_test.go [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+var (
+	expiration = time.Now().Add(time.Hour)
+)
+
+// TestCircuitBreaker verifies that the circuit breaker prevents a nested Authz
+// message that contains a MsgTryUpgrade if the MsgTryUpgrade is not supported
+// in the current version.
+func TestCircuitBreaker(t *testing.T) {
+	config := encoding.MakeConfig(app.ModuleEncodingRegisters...)
+	testApp, keyRing := util.SetupTestAppWithGenesisValSet(app.DefaultInitialConsensusParams(), granter, grantee)
```
Comment (ninabarbakadze) [inline app/test/circuit_breaker_test.go:42]:
```suggestion
	testApp, keyRing := testutil.SetupTestAppWithGenesisValSet(app.DefaultInitialConsensusParams(), granter, grantee)
```

Comment (rootulp) [reply app/test/circuit_breaker_test.go:39]: Thanks for flagging, fixed.

File: app/test/circuit_breaker_test.go (hunk 2 above)
Comment (rootulp) [reply app/test/circuit_breaker_test.go:11]: Thanks for flagging, fixed.

**Review by ninabarbakadze — APPROVED** [review]

File: app/test/circuit_breaker_test.go [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+	_, err = testApp.AuthzKeeper.Grant(ctx, msg)
+	assert.Error(t, err)
+	assert.ErrorContains(t, err, "/celestia.signal.v1.Msg/TryUpgrade doesn't exist.: invalid type")
+
+	testApp.BeginBlocker(ctx, abci.RequestBeginBlock{Header: header})
+
+	tryUpgradeTx := newTryUpgradeTx(t, signer, granterAddress)
+	res := testApp.DeliverTx(abci.RequestDeliverTx{Tx: tryUpgradeTx})
+	assert.Equal(t, uint32(0x25), res.Code, res.Log)
+	assert.Contains(t, res.Log, "message type /celestia.signal.v1.MsgTryUpgrade is not supported in version 1: feature not supported")
```
Comment (cmwaters) [inline app/test/circuit_breaker_test.go:61]: I didn't realise it called the antehandler again in `DeliverTx`

**Review by cmwaters — APPROVED** [review]
LGTM

**Pull request merged**